### ISSUE TITLE
Expressions can be typed as open generic types

### ISSUE THREAD

**Issue by JonHanna** [dotnet/corefx#8081]
The likes of `Expression.Default(typeof(List<>))`, `Expression.Constant(null, typeof(Dictionary<,>))` and some other expressions (though often other rules will get in the way) are allowed.

Since these represent a value of an open generic type, they aren't really meaningful.

They can have a value of null which can be used as long as it is converted to `object` and both compilation types accept `Expression.Lambda<Func‎<object>>(Expression.Default(typeof(List<>)))` as an expression that can be compiled into a delegate that returns null.

While they can be pressed into service, it seems more appropriate to prohibit them.

Factory methods that have at least one overload that takes `Type` arguments directly. Will check these off when they're either found to already prohibit such types, or fixed to do so (PRs just adding tests of already-correct behaviour marked immediately, PRs fixing behaviour upon acceptance):

- [x] `Block` (entailed in check that type is reference-assignable from type of last child expression).
- [x] `Break`
- [x] `Call` #17333
- [x] `Catch`
- [x] `Condition`
- [x] `Constant`
- [x] `Continue`
- [x] `Convert`
- [x] `ConvertChecked`
- [x] `Default`
- [x] `Dynamic` (Entailed in check that arguments match delegate method).
- [x] `Field` #16231
- [x] `GetActionType` (Exception: Has practical use here)
- [x] `GetDelegateType` (Exception: Has practical use here)
- [x] `GetFuncType` (Exception: Has practical use here)
- [x] `Goto`
- [x] `Label` #16140
- [x] `Lambda` #16233
- [x] `MakeCatchBlock`
- [x] `MakeDynamic` (Entailed in check that arguments match delegate method).
- [x] `MakeGoto`
- [x] `MakeTry` (entailed by check that type is void or that body and catch types are reference-assignable)
- [x] `MakeUnary` Waiting on #16141
- [x] `New`
- [x] `NewArrayBounds`
- [x] `NewArrayInit`
- [x] `Parameter` (byref types allowed)
- [x] `Property` #16231
- [x] `Rethrow` #16141
- [x] `Return`
- [x] `Switch` (Entailed by check that type is void or default and case types are reference-assignable)
- [x] `Throw` #16141
- [x] `TryGetActionType` (Exception: Has practical use here)
- [x] `TryGetFuncType` (Exception: Has practical use here)
- [x] `TypeAs`
- [x] `TypeEqual` #16150 (pointer types allowed; always results in `false`)
- [x] `TypeIs` #16150 (pointer types allowed; always results in `false`)
- [x] `Unbox` #16186
- [x] `Variable`

Also the following have overrides that take `MemberInfo`, which could be generic and/or be related to an open generic type. Most will block this with restrictions they place on those `MemberInfo` arguments relating to other types, but they need to be checked.

- [x] `Add` #16142
- [x] `AddAsign` #16142
- [x] `AddAsignChecked` #16142
- [x] `AddChecked` #16142
- [x] `And` #16142
- [x] `AndAlso` #16142
- [x] `AndAssign` #16142
- [ ] `Bind` #18610
- [x] `Call`
- [x] `Convert` #16437
- [x] `ConvertChecked` #16437
- [x] `Decrement` #16437
- [x] `Divide` #16142
- [x] `DivideAssign` #16142
- [x] `ElementInit`
- [x] `Equal` #16142
- [x] `ExclusiveOr` #16142
- [x] `ExclusiveOrAssign` #16142
- [x] `Field` #16231
- [x] `GreaterThan` #16142
- [x] `GreaterThanOrEqual` #16142
- [x] `Increment` #16437
- [x] `IsFalse` #16437
- [x] `IsTrue` #16437
- [x] `LeftShift` #16142
- [x] `LeftShiftAssign` #16142
- [x] `LessThan` #16142
- [x] `LessThanOrEqual` #16142
- [x] `ListBind` #18011
- [x] `ListInit` #16624
- [x] `MakeBinary` #16142
- [x] `MakeIndex`
- [x] `MakeMemberAccess`
- [x] `MakeUnary` #16437
- [x] `MemberBind` (Covered by validation added in  #18011 but also entailed by need of type to match).
- [x] `Modulo` #16142
- [x] `ModuloAssign` #16142
- [x] `Multiply` #16142
- [x] `MultiplyAssign` #16142
- [x] `MultiplyAssignChecked` #16142
- [x] `MultiplyChecked` #16142
- [x] `Negate` #16437
- [x] `NegateChecked` #16437
- [x] `New`
- [x] `Not` #16437
- [x] `NotEqual` #16142
- [x] `OnesComplement` #16437
- [x] `Or` #16142
- [x] `OrAssign` #16142
- [x] `OrElse` #16142
- [x] `PostDecrementAssign` #16437
- [x] `PostIncrementAssign` #16437
- [x] `Power` #16142
- [x] `PowerAssign` #16142
- [x] `PreDecrementAssign` #16437
- [x] `PreIncrementAssign` #16437
- [x] `Property` #16231
- [x] `RightShift` #16142
- [x] `RightShiftAssign` #16142
- [x] `Subtract` #16142
- [x] `SubtractAssign` #16142
- [x] `SubtractAssignChecked` #16142
- [x] `SubtractChecked` #16142
- [x] `Switch` #16344
- [x] `UnaryPlus` #16437

**Comment by JonHanna** [comment]
Several of them can also be typed as byref and/or pointer types.

**Comment by hughbe** [comment]
Should the following be prohibited or allowed (they are currently allowed):
```
Expression.Variable(typeof(GenericClass<>);
Expression.Parameter(typeof(GenericClass<>);
```

I'm fairly sure the Variable one is a bug (the C# compiler doesn't allow it), but I'm uncertain about Parameter.

**Comment by svick** [comment]
@hughbe Since `Variable()` and `Parameter()` are almost interchangeable, I think they should behave the same here.

And as far as open generic types go, I don't see any reason to allow them.

**Comment by hughbe** [comment]
@svick perf, I'll make a PR

**Comment by JonHanna** [comment]
I've just found that `TypeAs` can be typed as pointer and ref types, so there's at least one case that is still affected by this.

**Comment by hughbe** [comment]
Oh whoops I'm not sure I meant to close this issue. Feel free to reopen if u want

**Comment by JonHanna** [comment]
I can't due to https://github.com/isaacs/github/issues/583

**Comment by JonHanna** [comment]
I had been working systematically on this, before I wasn't able to contribute much for a while last summer. Need to track what, if anything remains in this, issue. Shall update the opening comment with a task list.

**Comment by JonHanna** [comment]
Marking `MakeIndex` and `MakeMemberAccess` as okay as they make quite simple calls into methods already covered.
